src/subscriptions.rs

```diff
@@ -1,41 +1,316 @@
+}
+
+#[derive(Clone, Debug, Decode)]
+struct GrandpaJustification {
+	pub round: u64,
+	pub commit: Commit,
+	pub _votes_ancestries: Vec<Header>,
+}
+
+impl<'de> Deserialize<'de> for GrandpaJustification {
+	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+	where
+		D: serde::Deserializer<'de>,
+	{
+		let encoded = bytes::deserialize(deserializer)?;
+		Self::decode(&mut &encoded[..])
+			.map_err(|codec_err| D::Error::custom(format!("Invalid decoding: {:?}", codec_err)))
+	}
+}
+
+#[derive(Clone, Debug)]
+enum Messages {
+	Justification(GrandpaJustification),
+	ValidatorSetChange((Vec<EdPublic>, u64)),
+	NewHeader(Header, Instant),
+}
+
+async fn subscribe_check_and_process(
+	subxt_client: OnlineClient<AvailConfig>,
+	message_tx: Sender<(Header, Instant)>,
+) -> Result<()> {
+	let mut header_subscription = subxt_client
+		.rpc()
+		.subscribe_finalized_block_headers()
+		.await?;
+
+	// Current set of authorities, implicitly trusted, fetched from grandpa runtime.
+	let grandpa_valset_raw = subxt_client
+		.runtime_api()
+		.at(None)
+		.await?
+		.call_raw("GrandpaApi_grandpa_authorities", None)
+		.await?;
+
+	// Decode result to proper type - ed25519 public key and u64 weight.
+	let grandpa_valset: Vec<(EdPublic, u64)> = Decode::decode(&mut &grandpa_valset_raw[..])?;
+
+	// Drop weights, as they are not currently used.
+	let mut validator_set: Vec<EdPublic> = grandpa_valset.iter().map(|e| e.0).collect();
+
+	// Set ID is acquired in a separate storage query. It is necessary, because it is a part of message being signed.
+
+	// Form a query key for storage
+	let set_id_key = api::storage().grandpa().current_set_id();
+	// Fetch the set ID from storage at current height
+	let mut set_id = subxt_client
+		.storage()
+		// None means current height
+		.at(None)
+		.await?
+		.fetch(&set_id_key)
+		.await?
+		.expect("The set_id shold exist");
+
+	// Get last (implicitly trusted) finalized block number
+	let last_finalized_block_hash = subxt_client.rpc().finalized_head().await?;
+	let mut last_finalized_block_header = subxt_client
+		.rpc()
+		.header(Some(last_finalized_block_hash))
+		.await?
+		.expect("Finalized head should exist");
+
+	debug!("Current set: {:?}", validator_set);
+
+	// Forming a channel for sending any relevant events gathered asynchronously through Substrate WS API.
+	let (msg_sender, mut msg_receiver) = unbounded_channel::<Messages>();
+
+	// Task that produces headers and new validator sets
+	tokio::spawn({
+		let subxt_client = subxt_client.clone();
+		let msg_sender = msg_sender.clone();
+		async move {
+			while let Some(Ok(header)) = header_subscription.next().await {
+				let received_at = Instant::now();
+				let head_hash: H256 = Encode::using_encoded(&header, blake2_256).into();
+				msg_sender
+					.send(Messages::NewHeader(header, received_at))
+					.expect("Receiver should not be dropped.");
+
+				// Fetch all events at the incoming header hight.
+				let events = subxt_client
```

Comment:
The total events per block can be quite big. Is there a more optimized way of fetching pallet based events?

Reply:
I haven't found a way, so far. All the other API's either supply authority keys in a different format, requiring more RPC's or aren't working the way they should.